### DISCUSSION TITLE
Is there a way to plot ad-stock curves? Has anyone tried it?

### DISCUSSION THREAD

**Discussion by AruunnAnanda** [google/lightweight_mmm#174] [Q&A]
I'm interested in knowing how the ad-stock curves look like for each of the marketing channel. Is there a way to extract and plot it?

Thanks in advance for your help!

**Comment by steven-struglia** [comment]
When you say the adstock curves, do you mean the regular spend over time vs. adstocked spend over time? Or do you mean the percentage of realized spend across time? I have both, so let me share here.


![image](https://user-images.githubusercontent.com/40417554/220172731-b24c8e4d-d26b-4afb-acdd-4178a235ce02.png)
For the above graph (regular spend vs. adstocked spend) we have the following code, where `mmm` is the media mix model that you have fit already and gathered parameters from. `mdsp_cols` or `media_cols` (better variable name to use) is simply the set of any media channel feature columns that you have (i.e. facebook, tiktok, twitter, etc.).

```
def plot_adstock(mmm):
    
    def adstock(data, lag_weight):
        adstocked = []
        adstock_val = 0
        for x in data:
            new_adstock = x + adstock_val*lag_weight
            adstock_val = new_adstock
            adstocked.append(new_adstock)
        return adstocked
    
    lag_weights = mmm.trace['lag_weight'].T
    for x,col in list(zip(lag_weights, mdsp_cols)):
        x = x.mean()
        adstocked = adstock(np.array(data[col]),x)
        plt.plot(np.array(data[col]), label=f'{col}')
        plt.plot(adstocked, label=f'adstock_{col}')
        plt.legend()
        plt.show()
        
plot_adstock(mmm)
```


![image](https://user-images.githubusercontent.com/40417554/220173498-e6865134-d82b-4128-97af-838b2f3cc73d.png)

For a time-shift curve (as we call it) that represents the percentage of realized spending over each day from the original spend date, we can visualize what percentage of our spending on an ad channel at date `X_0` will carry over into date `X_0`+1, as well as how much spending impact we have on any day `X_0` + `n_days`. We have the following code:

```
n_days = 180
lag_weights = mmm.trace['lag_weight'].T
shift_curves = pd.DataFrame()
for x,col in list(zip(lag_weights, mdsp_cols)):
    x = x.mean()
    # Since lag_weight is the fraction we carry into the next day, 1-lag_weight gives us the realized spend
    #      day 0                  day 1                               day 2                                          day 3
    # (1-lag_weight) + (lag_weight)(1-lag_weight)  + (lag_weight)(lag_weight)(1-lag_weight) + (lag_weight)(lag_weight)(lag_weight)(1-lag_weight) + .....
    lag_weight = x
    realized = 1-x
    shift_vals = [realized]
    for i in range(1,n_days):
        shift_vals.append(shift_vals[-1] + realized*(lag_weight**i))
    shift_curves[col] = shift_vals
```

**Reply by AruunnAnanda** [reply]
I was referring to the ad-stocked spend over time. Many thanks for sharing the code! @steven-struglia .

**Comment by hellojungkook** [comment]
this is super helpful. thanks!
